### ISSUE TITLE
user-defined constant G

### ISSUE THREAD

**Issue by jinz2014** [intel/llvm#11335]
a user-define constant 

#define G 9.81

seems to cause many compile errors

oneapi-2023.2.0/compiler/2023.2.0/linux/bin-llvm/../include/sycl/group.hpp:69:43: error: expected ')'
  private_memory(const group<Dimensions> &G)

**Comment by AlexeySachkov** [comment]
Hi @jinz2014,

Yes, that is a known issue, see #3677

We have made a few passes over the headers to fix reported conflicts, but haven't yet refactored them to be completely conflict-free and haven't added any testing or development guidelines for this

**Comment by jinz2014** [comment]
If it will never be conflict-free, please add some guidelines for users.  Thanks.

**Comment by jzc** [comment]
Perhaps one thing we could try to enforce throughout the headers is no use of all capital identifiers? Under standard naming conventions, macros are all capitals, and for variables names, capital variable names really only happen in a case like this, with a single capital letter, or maybe under the case of an abbreviation of the type of the parameter (e.g. in LLVM codebase, `const FunctionType *FT`). If we do want to implement this, I can look into writing a clang-tidy plugin for enforcing this.

**Comment by jinz2014** [comment]
It is good to search all identifiers (e.g. G, N, M) that are used in the header files ?  Then, these identifiers should not be used by users. 

```
template<class identifier> declaration;
template<typename identifier> declaration;
```

**Comment by AlexeySachkov** [comment]
> Perhaps one thing we could try to enforce throughout the headers is no use of all capital identifiers? Under standard naming conventions, macros are all capitals, and for variables names, capital variable names really only happen in a case like this, with a single capital letter, or maybe under the case of an abbreviation of the type of the parameter (e.g. in LLVM codebase, `const FunctionType *FT`). If we do want to implement this, I can look into writing a clang-tidy plugin for enforcing this.

Those have probably the biggest impact on users - based on reports I've seen. However, these are only a sub-set of all unreserved symbols we use. Writing a `tidy` plugin would be a must if we decide to enforce the policy about not using unreserved symbols, but we need to discuss this first:
- what would be a naming convention: `_G`, `__g`, `g`?
- are we in agreement to enforce a _partial_ policy, i.e. only prohibit all-capital identifiers, or are we going all-in and actually making our headers conflict-free?

@jzc, let's maybe gather opinions from the team, stakeholders/code owners and discuss the topic on some of our forums?

**Comment by cperkinsintel** [comment]
Personally, I think users defining single capital letters should simply be "known incompatibility" and end of story.  I don't think we should make any effort to fix this. 

  Who are these people doing `#define G` or `#define T` in modern C++?  It's a terrible practice that causes problems with many many other libraries, not just ours.  
  
  If the user wants to `#define G` then they can do so AFTER their inclusions are done.

**Comment by illuhad** [comment]
> Personally, I think users defining single capital letters should simply be "known incompatibility" and end of story. I don't think we should make any effort to fix this.
> Who are these people doing #define G or #define T in modern C++? It's a terrible practice that causes problems with many many other libraries, not just ours.
> If the user wants to #define G then they can do so AFTER their inclusions are done.

This. If it helps you as another data point, what you say also reflects our handling of such cases in the AdaptiveCpp implementation. We have zero intent of supporting such bad practices, and any such code is likely to break anyway - either due to the SYCL implementation, or any other C++ library. So even if you fixed this in your implementation, such SYCL code still would not be portable to other implementations like ours.
But I admit that for us it is perhaps a little easier to tell users that their bad code is the problem, because we don't have commercial support contracts.

**Comment by Pennycook** [comment]
> Personally, I think users defining single capital letters should simply be "known incompatibility" and end of story. I don't think we should make any effort to fix this.
> 
> Who are these people doing `#define G` or `#define T` in modern C++? It's a terrible practice that causes problems with many many other libraries, not just ours.

I disagree with this.  If we fix this, we fix it for all affected users.  If we don't fix this, everybody affected by this has to fix it themselves.  As far as I can tell, such macros exist because of some combination of the following:

- The code existed before the C++/SYCL version was written.
- This hasn't been a problem for competing programming models (e.g., OpenMP, CUDA).
- Other header-only C++ libraries (e.g., the STL) go to great lengths to avoid using reserved identifiers.
- There are rules defining identifiers that are reserved for use in implementation headers; anything that isn't reserved is expected to be available to developers.

> If the user wants to `#define G` then they can do so AFTER their inclusions are done.

I don't think this is a good solution.  A developer might encounter `#define G` in a header that is maintained by somebody else in their team, and so they don't even know the macro exists.  Why make our headers sensitive to include order if they needn't be?

And what if somebody has structured their code such that they define `G` on the command-line?  Making them change the way their code has always worked just to use our headers doesn't seem like a good user experience to me, and is likely to result in developers giving up and trying something else.

**Comment by Pennycook** [comment]
> Writing a `tidy` plugin would be a must if we decide to enforce the policy about not using unreserved symbols, but we need to discuss this first:

@AlexeySachkov I've looked into this, and there's already a clang-tidy plugin that could be used for this. 

[`bugprone-reserved-identifier`](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/reserved-identifier.html) can be used to detect the use of reserved identifiers.  But its behavior can also be inverted so that it complains about the use of any unreserved identifier instead.  It even supports a `--fix` option to automatically rename all the variables.

The only problem I see is that the default behavior is too aggressive.  For example, I ran it on this test:
```c++
namespace sycl {
template <int Dimensions> class group;
inline namespace _V1 {
namespace detail {
class Builder;

template <typename T, int Dimensions = 1> class private_memory {
public:
  private_memory(const group<Dimensions> &_G) { (void)G; }
};

} // namespace detail
}
} // namespace sycl
```
and it produced:
```c++
namespace __sycl {
template <int _Dimensions> class __group;
inline namespace _V1 {
namespace __detail {
class _Builder;

template <typename _T, int _Dimensions = 1> class __private_memory {
public:
  __private_memory(const group<_Dimensions> &_G) { (void)_G; }
};

} // namespace __detail
}
} // namespace __sycl
```

To fix this, somebody will need to go through the SYCL specification and build a list of names that our headers are _supposed_ to define.  Once we have that, we can provide this as a list of allowed identifiers:
```yaml
Checks: 'bugprone-reserved-identifier'
CheckOptions:
    - key: bugprone-reserved-identifier.Invert
      value: 1
    - key: bugprone-reserved-identifier.AllowedIdentifiers
      value: "sycl;private_memory;group;detail"
FormatStyle: 'file'
```

...which produces what we want:
```c++
namespace sycl {
template <int _Dimensions> class group;
inline namespace _V1 {
namespace detail {
class _Builder;

template <typename _T, int _Dimensions = 1> class private_memory {
public:
  private_memory(const group<_Dimensions> &_G) { (void)_G; }
};

} // namespace detail
}
} // namespace sycl
```

**Comment by jinz2014** [comment]
Thank you for the discussion.
 
Workaround (no source changes): hide the macro across the SYCL include

Use #pragma push_macro / #undef / #pragma pop_macro so G is invisible only while the SYCL headers are parsed, then restored for your own
  code:

  #define G 9.81
  #pragma push_macro("G")
  #undef G
  #include <sycl/sycl.hpp>
  #pragma pop_macro("G")
  int main() {
    double x = G;   // still 9.81 here
  }
